server/routes/api.js

```diff
@@ -1073,129 +1089,359 @@ router.post('/capcodeImport', isLoggedIn, function (req, res, next) {
+            res.json(defaults);
+          }
+        })
+        .catch((err) => {
+          logger.main.error(err);
+          return next(err);
+        })
+    }
+  })
+  .post(isAdmin, function (req, res, next) {
+    var id = req.params.id || req.body.id || null;
+    if (id == 'deleteMultiple') {
+      // do delete multiple
+      var idList = req.body.deleteList || [0, 0];
+      if (!idList.some(isNaN)) {
+        logger.main.info('Deleting: ' + idList);
+        db.from('users')
+          .del()
+          .where('id', 'in', idList)
+          .then((result) => {
+            res.status(200).send({ 'status': 'ok' });
+
+          }).catch((err) => {
+            res.status(500).send(err);
+          })
       } else {
-        throw 'Error parasing CSV header'
+        res.status(500).send({ 'status': 'id list contained non-numbers' });
       }
-    })
-    .catch((err) => {
-      res.status(500).send(err)
-      logger.main.error(err)
-    })
-});
+    } else {
+      if (req.body.username && req.body.email && req.body.givenname) {
+        if (id == 'new') {
+          id = null;
+        }
+        var username = req.body.username;
+        var givenname = req.body.givenname;
+        var surname = req.body.surname || '';
+        var email = req.body.email;
+        var password = req.body.newpassword || null;
+        var role = req.body.role || 'user';
+        var status = req.body.status || 'disabled';
+        var lastlogondate = Date.now()
+        console.time('insert');
+        db.from('users')
+          .returning('id')
+          .where('id', '=', id)
+          .modify(function (queryBuilder) {
+            if (password != null) {
+              const salt = bcrypt.genSaltSync();
+              const hash = bcrypt.hashSync(password, salt);
+              if (id == null) {
+                queryBuilder.insert({
+                  id: id,
+                  username: username,
+                  givenname: givenname,
+                  surname: surname,
+                  password: hash,
+                  email: email,
+                  role: role,
+                  status: status,
+                  lastlogondate: lastlogondate
+                })
+              } else {
+                queryBuilder.update({
+                  id: id,
+                  username: username,
+                  givenname: givenname,
+                  surname: surname,
+                  password: hash,
+                  email: email,
+                  role: role,
+                  status: status,
+                  lastlogondate: lastlogondate
+                })
+              }
+            } else {
+              queryBuilder.update({
+                id: id,
+                username: username,
+                givenname: givenname,
+                surname: surname,
+                email: email,
+                role: role,
+                status: status,
+                lastlogondate: lastlogondate
+              })
```

Comment:
Feel like this chunk could be simplified a lot - build an object and then update or insert that object?

Reply:
Maybe objection can help simplify this part